File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/Range.java
package pl.allegro.experiments.chi.chiserver.domain.experiments;

import com.google.common.base.Preconditions;

import java.util.Objects;

abstract public class Range {
    private final int from;
    private final int to;

    Range(int from, int to) {
        Preconditions.checkArgument(from >= 0, "Range.from < 0");
        Preconditions.checkArgument(to > from, "Range: to should be > from, got " + from, " " + to);
        this.from = from;
        this.to = to;
    }

    public int getFrom() {
        return from;
    }

    public int getTo() {
        return to;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || !(o instanceof Range)) return false;
        Range range = (Range) o;
        return from == range.from && to == range.to;
    }

    @Override
    public int hashCode() {
        return Objects.hash(from, to);
    }

    @Override
    public String toString() {
        return "Range{" +
                "from=" + from +
                ", to=" + to +
                '}';
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/experiments/Range.kt
package pl.allegro.experiments.chi.chiserver.domain.experiments


/**
 * Represents range of numbers, right-open.
 * Left bound is inclusive, right bound is NOT inclusive.
 */
abstract class Range<out T : Number>(val from: T, val to: T) {

    init {
        require(from.toDouble() >= 0) { "Range.from < 0 in Range($from..$to)" }
        require(from.toDouble() <= to.toDouble()) { "Range.from > Range.to in Range($from..$to)" }
    }
}

class PercentageRange(from: Int, to: Int) : Range<Int>(from, to) {
    init {
        require(to <= 100) { "Range.to > 100 in PercentageRange($from..$to)" }
    }
}
